src/stake-pool/utils.ts

```diff
@@ -712,13 +857,51 @@ function calcWithdrawalReceipt(
   const lamportsReceived = Numberu64.cloneFromBN(
     num.add(poolTokenSupply).sub(new Numberu64(1)).div(poolTokenSupply),
   );
+
   return {
     dropletsUnstaked: Numberu64.cloneFromBN(dropletsToUnstake),
     lamportsReceived,
     dropletsFeePaid,
   };
 }
 
+function calcWithdrawalReceiptInverse(
+  lamportsToReceive: Numberu64,
+  stakePool: schema.StakePool,
+): WithdrawalReceipt {
+  const { withdrawalFee, totalStakeLamports, poolTokenSupply } = stakePool;
+
+  // If poolTokenSupply == 0, then below values should all be 0 as well.
+
+  const dropletsBurnt = lamportsToReceive
+    .mul(poolTokenSupply)
+    .div(totalStakeLamports);
+
+  const dropletsToUnstake = dropletsBurnt
+    .mul(withdrawalFee.denominator)
+    .div(
+      Numberu64.cloneFromBN(
+        withdrawalFee.denominator.sub(withdrawalFee.numerator),
+      ),
+    );
+
+  const hasFee =
+    !withdrawalFee.numerator.isZero() && !withdrawalFee.denominator.isZero();
+  const dropletsFeePaid = hasFee
+    ? Numberu64.cloneFromBN(
+        withdrawalFee.numerator
+          .mul(dropletsToUnstake)
+          .div(withdrawalFee.denominator),
+      )
+    : new Numberu64(0);
```

Comment:
think you need to move this part above the `const dropletsToUnstake = ...` line and perform a similar check for it too else it might divide by 0

Reply:
but `dropletsFeePaid` is dependent on `dropletsToUnstake`, so how would I define that before?

Reply:
sorry i meant specifically the `const hasFee = !withdrawalFee.numerator.isZero() && !withdrawalFee.denominator.isZero();` line and then doing a similar `hasFee ? ` ternary check for `dropletsToUnstake`